lib/racecar/forking_runner.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+module Racecar
+  class ForkingRunner
+    def initialize(runner:, config:, logger:, liveness_monitor: LivenessMonitor.new)
+      @runner = runner
+      @config = config
+      @logger = logger
+      @pids = []
+      @liveness_monitor = liveness_monitor
+      @running = false
+    end
+
+    attr_reader :config, :runner, :logger, :pids, :liveness_monitor
+    private     :config, :runner, :logger, :pids, :liveness_monitor
+
+    def run
+      config.prefork.call
+      @running = true
+
+      @pids = config.forks.times.map do |n|
+        pid = fork do
+          liveness_monitor.child_post_fork
+          config.postfork.call
+
+          liveness_monitor.on_exit do
+            logger.warn "Supervisor dead, exiting."
+            runner.stop
+          end
+
+          runner.run
+        end
+        logger.info "Racecar forked consumer process #{pid}"
+
+        pid
+      end
+
+      liveness_monitor.parent_post_fork
+
+      install_signal_handlers
+
+      wait_for_child_processes
+    end
+
+    def stop
+      @running = false
+      $stdout.puts "Racecar::ForkingRunner runner stopping #{Process.pid}"
+      terminate_workers
+    end
+
+    def running?
+      !!@running
+    end
+
+    private
+
+    def terminate_workers
+      pids.each do |pid|
+        begin
+          Process.kill("TERM", pid)
+        rescue Errno::ESRCH
+        end
+      end
+    end
+
+    def check_workers
+      pids.each do |pid|
+        unless worker_running?(pid)
+          $stdout.puts("A forker worker has exited. Shuttin' it down ...")
+          stop
```

Comment:
Yes, it would unexpected for this to happen, so rather than try to recover from a potentially bad state the whole thing shuts down so Kubernetes can bring up a new replica.

I'm not against trying to recover and replace a lost worker process but that is quite a bit more complex.